File: src/lib/compress.ts
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { Datapoint, DatapointsCompressed } from "../types";

// See `compress.py` for compression code on the python side
export function uncompress(compressed_data: DatapointsCompressed) {
    const columns = compressed_data.columns;
    const rows = compressed_data.rows;
    return rows.map(function(row) {
        const values = {};
        var dp: Datapoint = {
            'uid': row[0],
            'from_uid': row[1],
            'values': values,
        };
        columns.forEach(function(column, i) {
            values[column] = row[i + 2];
        });
        return dp;
    });
 }

File: hiplot/compress.py
# Copyright (c) Facebook, Inc. and its affiliates.
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.


import typing as tp

from .experiment import Datapoint


def _build_columns_list(datapoints: tp.List[Datapoint]) -> tp.List[str]:
    columns: tp.Set[str] = set()
    for dp in datapoints:
        columns = columns.union(dp.values.keys())
    for reserved_col in ['uid', 'from_uid']:
        try:
            columns.remove(reserved_col)
        except KeyError:
            pass
    return list(columns)


def compress(datapoints: tp.List[Datapoint]) -> tp.Dict[str, tp.Any]:
    columns = _build_columns_list(datapoints)
    rows: tp.List[tp.Any] = []
    for dp in datapoints:
        d: tp.List[tp.Any] = [dp.uid, dp.from_uid]
        for c in columns:
            d.append(dp.values.get(c))
        rows.append(d)
    return {
        "columns": columns,
        "rows": rows
    }
